crates/integrations/src/datafusion/schema.rs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{any::Any, sync::Arc};
+
+use dashmap::DashMap;
+use datafusion::{
+    catalog::schema::SchemaProvider, datasource::TableProvider, error::DataFusionError,
+};
+use futures::FutureExt;
+use iceberg::{Catalog, NamespaceIdent, Result};
+
+use crate::datafusion::table::IcebergTableProvider;
+
+pub(crate) struct IcebergSchemaProvider {
+    tables: DashMap<String, Arc<dyn TableProvider>>,
+}
+
+impl IcebergSchemaProvider {
+    pub(crate) async fn try_new(
+        client: Arc<dyn Catalog>,
+        namespace: &NamespaceIdent,
+    ) -> Result<Self> {
+        let table_names: Vec<String> = client
+            .list_tables(namespace)
+            .await?
+            .iter()
+            .map(|t| t.name().to_owned())
+            .collect();
+
+        let mut tables = Vec::new();
+        for name in table_names {
+            let provider = IcebergTableProvider::try_new(client.clone(), namespace, &name).await?;
+            let provider = Arc::new(provider) as Arc<dyn TableProvider>;
+
+            tables.push((name, provider));
+        }
+
+        Ok(IcebergSchemaProvider {
+            tables: tables.into_iter().collect(),
+        })
+    }
+}
+
+impl SchemaProvider for IcebergSchemaProvider {
```

Comment:
I realize these are just stubs, I'd be happy to help fill in the guts here and for the `TableProvider` traits.

Reply:
yes - I just wanted to share a draft as early as possible - in order to get the design ideas aligned. And sure - any help is appreciated since I'm not so familiar with the Datafusion traits and I think we also need to implement `ExecutionPlan` in order to provide the correct return type for the `TableProvider`...?

Reply:
Sweet, happy to help... for the catalog data [delta-rs](https://github.com/delta-io/delta-rs/blob/5eade5e1f07e80ecd053bd538e005894b7e88a8f/crates/core/src/data_catalog/unity/datafusion.rs#L140) is probably a good example to crib from. Also, using `async_trait` will improve the aesthetics of `IcebergSchemaProvider::table` and `IcebergTableProvider::scan`.

For execution then (`ExecutionPlan`), to your point it needs to be implemented. Similarly [`DeltaTable`](https://github.com/delta-io/delta-rs/blob/3094bd28cee94b41b502cff6a479965d1e81e990/crates/core/src/delta_datafusion/mod.rs#L654-L705) implements `TableProvider` which uses [`DeltaScan`](https://github.com/delta-io/delta-rs/blob/3094bd28cee94b41b502cff6a479965d1e81e990/crates/core/src/delta_datafusion/mod.rs#L789-L798) that implements `ExecutionPlan`.

What makes sense to you help-wise, so I don't step on your toes? Maybe I could branch off this and take a crack at `IcebergScan` or just commit to this branch or something else?

Reply:
you can simply pick what you like - and let me know, how you'd want to contribute - if you want to commit to this branch, I guess I need to invite you to my fork?

Reply:
Ok great, I'll get started on an `IcebergScan` implementation, or at least plan out some work in more detail and follow up if there's something ambiguous.

Reply:
I looked a little more at this, and it seems reasonable to get a first version up and running. Blocker right now is https://github.com/apache/iceberg-rust/pull/277 like you mentioned.